20240627-BC-listapi.md

```diff
@@ -0,0 +1,159 @@
+}
+```
+
+### Default values
+
+- Prefix: “”
+- Sorting: “default”
+- Page limit: 50
+- Next token: “”
+
+## **Pagination**
+
+The two most common pagination strategies are token and offset-based pagination.
+
+**Offset-based pagination** 
+Uses a fixed offset and limit to retrieve a subset of results from a larger dataset. This method is common in relational databases and is implemented with the `LIMIT` and `OFFSET` clauses. 
+
+It’s not common in no-sql databases, but it is very common in SQL databases. It relies on a table scan and skipping results until it reaches the offset value.
+
+**Token-based pagination** 
+
+Relies on a token usually equal to, or derived from the last element in the last returned page. 
+
+Very common in no-sql databases that do a scan across the keyspace. 
+
+In relational databases this method relies on an indexed column, such as a timestamp or an ID, to ensure efficient sorting and querying. For example:
+
+```bash
+SELECT * FROM items WHERE key > last_key_id ORDER BY key;
+```
+
+---
+
+Most often, offset-based pagination is not possible in no-sql databases, while it’s easy (even preferable) to implement in relational databases, so this proposal suggests using **token-based pagination** in the List API.
+
+Based on this decision, listing items will only be available forwards, and not backwards. To list previous pages, the application would have to keep track of the page tokens.
+
+## **Sorting**
+
+Sorting is required for token-based pagination in relational databases, so we must have a default sorting order.
+
+Some no-sql databases (ex. Azure blob store) don’t support sorting in descending order and others don’t support any sorting at all (ex. Redis). In these cases, we want to return an explicit error instead of failing silently.
+
+This might be restricting for use cases where the underlying state store needs to be swapped though. For example a team could use Redis for local development, and Postgres in production, and they wouldn’t be able to use the same application code, because the sorting clause would error on Redis, but pass on Postgres. That’s why we’re introducing the `Default` sorting option which will sort in ascending order for all databases that support it, and leave results unsorted for the databases that don’t.
+
+## SDKs
+
+All supported SDKs should be updated to implement the List API. SDKs should offer the option to fetch batch values of the returned keys.
+
+## Default behaviour for state stores with missing features
+
+Some of the state stores Dapr supports don’t provide the necessary capabilities for implementing the list API. For example, Memcached doesn’t provide a way to list keys,  Azure table storage can’t sort keys in descending order and so on. For those cases the list API will do a best effort to provide the closest functionality to the one defined in the API. The functionality will be specific to the data store and will be implemented on the component level.
```

Comment:
Different philosophies - but I no longer support this approach in Dapr building blocks and think we should instead have a new specialty building block.

Reply:
I like the idea of having specialty building blocks. I think the state store is doing too much. On the other hand, it also misses the List API, which is a basic operation. I think we should consider moving some state store components out of the building block into a new one and rename state store to key-value store. Another route is to deprecate the state store API and have multiple specialized building blocks. For the sake of people's time availability, the most realistic path is to rebrand state store to key/value store and add list API. Non-compliant components will partially implement the API until a specialized building block is created.